### PR TITLE
additional handling for non-standard DRS URIs [WA-383]

### PR DESCRIPTION
**Pull request by rsasch**
Closes https://broadworkbench.atlassian.net/browse/WA-383

### CODE REVIEW

File: cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala [hunk 1]
```diff
@@ -39,6 +39,8 @@ class DrsCloudNioFileProviderSpec extends AnyFlatSpecLike with Matchers with Moc
     fileSystemProvider.getScheme should be("drs")
     fileSystemProvider.getHost("drs://dg.123/abc") should be("dg.123")
     fileSystemProvider.getHost("drs://dg.example.com/abc") should be("dg.example.com")
+    fileSystemProvider.getHost("drs://dg.4503:dg.4503/abc") should be("dg.4503")
+    fileSystemProvider.getHost("drs://dg.4503:abc") should be("dg.4503")
```
Comment (kshakir) [inline cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala:43]: Are there more known-not-made-up-and-in-use [compact identifier-based](https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_calling_meta_resolver_apis_for_compact_identifier_based_drs_uris) (CIB) URIs that can be regression tested?

Comment (rsasch) [reply cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala:43]: I created these based on the examples in https://broadworkbench.atlassian.net/browse/WA-331.  Please suggest additional ones if I missed them.

File: cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileSystemProvider.scala [hunk 2]
```diff
@@ -42,16 +43,25 @@ class DrsCloudNioFileSystemProvider(rootConfig: Config,
     require(uriAsString.startsWith(s"$getScheme://"), s"Scheme does not match $getScheme")
 
     /*
-     * In some cases for a URI, the host name is null. For example, for DRS urls like 'drs://dg.123/123-123-123',
-     * even though 'dg.123' is a valid host, somehow since it does not conform to URI's standards, uri.getHost returns null. In such
-     * cases, authority is used instead of host. If there is no authority, use an empty string.
+     * In some cases, the URI will use a colon in a non-standard way, and therefore, we can't rely on
+     * the URI class to parse it.  Instead, we recognize it via a regular expression, and parse out the
+     * host name accordingly.
```
Comment (aednichols) [inline cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileSystemProvider.scala:48]:
```suggestion
     * DRS compact identifier URIs contain a string where conventional URIs have an integer
     * port (the `:8000` in `https://localhost:8000`). We can't rely on the URI class to parse it, 
     * so we recognize it via a regular expression, and parse out the host name accordingly.
```

File: cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala (hunk 1 above)
Comment (aednichols) [inline cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala:43]:
```suggestion
    fileSystemProvider.getHost("drs://dg.4503:dg.4503/abc") should be("dg.4503")
    fileSystemProvider.getHost("drs://dg.4503:abc") should be("dg.4503")
    fileSystemProvider.getHost("drs://:abc") should be(<fail gracefully>)
    fileSystemProvider.getHost("drs://dg.4503:") should be(<fail gracefully>)
```

**Review by aednichols — COMMENTED** [review]
The colon is technically not unexpected, so suggested a comment rewording

Also suggested some boundary condition tests

File: cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala (hunk 1 above)
Comment (rsasch) [reply cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala:43]: The `getHost` function, before these changes, would return an empty string if it couldn't find a valid host, so I've retained that behavior.  Does that count as failing gracefully?  If not, what benefit would be gained by throwing an error?

Comment (aednichols) [reply cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala:43]: Yes, that counts as failing gracefully.

Ungraceful would be if it threw a random exception.

Comment (aednichols) [reply cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala:43]: [TOL] Although it would be pretty cool to use `Option` and return `None` in that case

Comment (rsasch) [reply cloud-nio/cloud-nio-impl-drs/src/test/scala/cloud/nio/impl/drs/DrsCloudNioFileProviderSpec.scala:43]: Added the two tests, although `drs://dg.4503:` returns `dg.4503` as host, because that URI has a host, just no port or filepath.

**Review by aednichols — APPROVED** [review]
LGTM 👍

**Review by kshakir — APPROVED** [review]
